metrics/prometheus/prometheus.go

```diff
@@ -0,0 +1,302 @@
+	ExtraTTL    *PrometheusExtraTTLMetrics
+}
+
+type PrometheusRequestStatusMetric struct {
+	Duration      prometheus.Histogram
+	RequestStatus *prometheus.CounterVec
+}
+
+type PrometheusRequestStatusMetricByFormat struct {
+	Duration           prometheus.Histogram
+	PutBackendRequests *prometheus.CounterVec
+	RequestLength      prometheus.Histogram
+}
+
+type PrometheusConnectionMetrics struct {
+	ConnectionsOpened prometheus.Gauge
+	ConnectionsErrors *prometheus.CounterVec
+}
+
+type PrometheusExtraTTLMetrics struct {
+	ExtraTTLSeconds prometheus.Histogram
+}
+
+//	Init functions
+func CreatePrometheusMetrics(cfg config.PrometheusMetrics) *PrometheusMetrics {
+	cacheWriteTimeBuckets := []float64{0.001, 0.002, 0.005, 0.01, 0.025, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5, 1}
+	requestSizeBuckets := []float64{0.001, 0.002, 0.005, 0.01, 0.025, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5, 1} // TODO: tweak
+	registry := prometheus.NewRegistry()
+	promMetrics := &PrometheusMetrics{
+		Registry: registry,
+		Puts: &PrometheusRequestStatusMetric{
+			Duration: newHistogram(cfg, registry,
+				"puts_current_url_duration",
+				"Duration in seconds Prebid Cache takes to process put requests.",
+				cacheWriteTimeBuckets,
+			),
+			RequestStatus: newCounterVecWithLabels(cfg, registry,
+				"puts_current_url",
+				"Count of total requests to Prebid Server labeled by status.",
+				[]string{StatusKey},
+			),
+		},
+		Gets: &PrometheusRequestStatusMetric{
+			Duration: newHistogram(cfg, registry,
+				"gets_current_url_duration",
+				"Duration in seconds Prebid Cache takes to process get requests.",
+				cacheWriteTimeBuckets,
+			),
+			RequestStatus: newCounterVecWithLabels(cfg, registry,
+				"gets_current_url",
+				"Count of total get requests to Prebid Server labeled by status.",
+				[]string{StatusKey},
+			),
+		},
+		PutsBackend: &PrometheusRequestStatusMetricByFormat{
+			Duration: newHistogram(cfg, registry,
+				"puts_backend_duration",
+				"Duration in seconds Prebid Cache takes to process backend put requests.",
+				cacheWriteTimeBuckets,
+			),
+			PutBackendRequests: newCounterVecWithLabels(cfg, registry,
+				"puts_backend",
+				"Count of total requests to Prebid Cache labeled by format, status and whether or not it comes with TTL",
+				[]string{FormatKey},
+			),
+			RequestLength: newHistogram(cfg, registry,
+				"puts_backend_request_size_bytes",
+				"Size in bytes of a backend put request.",
+				requestSizeBuckets,
+			),
+		},
+		GetsBackend: &PrometheusRequestStatusMetric{
+			Duration: newHistogram(cfg, registry,
+				"gets_backend_duration",
+				"Duration in seconds Prebid Cache takes to process backend get requests.",
+				cacheWriteTimeBuckets,
+			),
+			RequestStatus: newCounterVecWithLabels(cfg, registry,
+				"gets_backend",
+				"Count of total backend get requests to Prebid Server labeled by status.",
+				[]string{StatusKey},
+			),
+		},
+		Connections: &PrometheusConnectionMetrics{
+			ConnectionsOpened: newGaugeMetric(cfg, registry,
```

Comment:
Please replace the connections guage with connections opened and connections closed counters as we have in Prebid Server. The value of the guage can be inferred by Prometheus given the raw data.

Reply:
done